### ISSUE TITLE
:bnext cursor positioning is different for a hidden buffer vs a buffer shown in another window

### ISSUE THREAD

**Issue by zeertzjq** [vim/vim#11558]
### Steps to reproduce

1. Run `vim --clean`
2. Source the following file:
```vim
set hidden
call setline(1, range(&lines + 10))
normal! G
vsplit
enew
bnext
```
3. Cursor in the left window is at the top of the file

### Expected behaviour

Cursor in the left window is at the bottom of the file. The is the behavior if I removed the `:vsplit`, and if I revert patch 8.1.0425 I also get this behavior.

### Version of Vim

9.0.813

### Environment

Operating system: Arch Linux
Terminal: kitty
Value of $TERM: xterm-kitty
Shell: fish

**Comment by zeertzjq** [comment]
I don't think this has anything to do with autocommands. `:bnext` still moves the cursor to last position when the buffer is not loaded even if I remove the autocommand:
```vim
augroup vimStartup | au! | augroup END
set hidden
call setline(1, range(&lines + 10))
normal! G
enew
bnext
```

And if revert patch 8.1.0425:
```diff
diff --git a/src/buffer.c b/src/buffer.c
index de4c40b58..78b41cd26 100644
--- a/src/buffer.c
+++ b/src/buffer.c
@@ -1906,9 +1906,6 @@ enter_buffer(buf_T *buf)
     // mark cursor position as being invalid
     curwin->w_valid = 0;
 
-    buflist_setfpos(curbuf, curwin, curbuf->b_last_cursor.lnum,
-					      curbuf->b_last_cursor.col, TRUE);
-
     // Make sure the buffer is loaded.
     if (curbuf->b_ml.ml_mfp == NULL)	// need to load the file
     {
```

The cursor is then also moved when the buffer is shown in another window even if I remove the autocommand:
```vim
augroup vimStartup | au! | augroup END
set hidden
call setline(1, range(&lines + 10))
normal! G
vsplit
enew
bnext
```

**Comment by zeertzjq** [comment]
Taking another look at patch 8.1.0425, it seems that fix doesn't always work. If the buffer isn't visible in any window, then `curwin` still cannot be set to a window that contains the buffer.

**Comment by brammool** [comment]
> Taking another look at patch 8.1.0425, it seems that fix doesn't
> always work. If the buffer isn't visible in any window, then `curwin`
> still cannot be set to a window that contains the buffer.

You mean that find_win_for_curbuf() may leave "curwin" as it was, and
then curwin->w_buffer differs from "curbuf"?  That is indeed dangerous.
Can you reproduce a failure with this situation?

We probably should use aucmd_prepbuf().  It adds some overhead, but it
was especially designed to handle changes in a buffer that isn't the
current one.  Before I make that change it would be good to have a test
for any failures with the current implementation.

After that we'll have to see if updating the last cursor position is
also fixed.

-- 
From "know your smileys":
 :~)	A man with a tape recorder up his nose

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by brammool** [comment]
I added the patch to make sure the buffer is changed in the context of a window.  Please check that either this has fixed the problem or it can still be reproduced.
If this did fix the problem, can you write a regression test?
If did this not fix the problem, what is the simplest way to reproduce?  Ideally in the form of a failing test.

**Comment by brammool** [comment]
Perhaps patch 9.0.0961 also matters for this issue.
Otherwise the questions from 8 days ago still matter.

**Comment by zeertzjq** [comment]
After over two years I decided to take another look at this issue and I think calling `buflist_setfpos()` here in `enter_buffer()` does not make sense. Just 30 lines below it restores the window-local last cursor position (`wi_fpos`):
```c
    // If autocommands did not change the cursor position, restore cursor lnum
    // and possibly cursor col.
    if (curwin->w_cursor.lnum == 1 && inindent(0))
	buflist_getfpos();
```
Therefore, it does not make sense to replace the window-local last cursor position (`wi_fpos`) with the buffer-local one (`b_last_cursor`) just before that.

**Comment by zeertzjq** [comment]
cc @h-east what do you think?

**Comment by zeertzjq** [comment]
Also, using `:bprev` to reproduce this is probably more portable:
```vim
call setline(1, range(&lines + 10))
normal! G
vsplit
enew
bprev
```

**Comment by h-east** [comment]
@zeertzjq 
I would have appreciated any information about why you ID-called me.  
I see - it's because I'm the author of the problematic patch (8.1.0425).

Give me a moment to remember and investigate.

**Comment by h-east** [comment]
I investigated.

The root cause of [8.1.0425](https://github.com/vim/vim/commit/9cea87c5775948a35098f3602746c20ecf95dbcd) has been fixed in [8.1.0786](https://github.com/vim/vim/commit/10772307c4e5299ed45470f92779f089a00d841e).
Therefore, 8.1.0425 should be reverted.

```
commit 10772307c4e5299ed45470f92779f089a00d841e (HEAD, tag: v8.1.0786)
Author: Bram Moolenaar <Bram@vim.org>
Date:   Sun Jan 20 18:25:54 2019 +0100

    patch 8.1.0786: ml_get error when updating the status line

    Problem:    ml_get error when updating the status line and a terminal had its
                scrollback cleared.  (Chris Patuzzo)
    Solution:   Check the cursor position when drawing the status line.
                (closes #3830)
```
